internal/datastore/proxy/replicated.go

```diff
@@ -0,0 +1,341 @@
+	if err := rr.determineSource(ctx); err != nil {
+		return nil, err
+	}
+
+	return rr.chosenReader.ListAllCaveats(ctx)
+}
+
+func (rr *checkingStableReader) LookupCaveatsWithNames(ctx context.Context, names []string) ([]datastore.RevisionedCaveat, error) {
+	if err := rr.determineSource(ctx); err != nil {
+		return nil, err
+	}
+
+	return rr.chosenReader.LookupCaveatsWithNames(ctx, names)
+}
+
+func (rr *checkingStableReader) QueryRelationships(
+	ctx context.Context,
+	filter datastore.RelationshipsFilter,
+	options ...options.QueryOptionsOption,
+) (datastore.RelationshipIterator, error) {
+	if err := rr.determineSource(ctx); err != nil {
+		return nil, err
+	}
+
+	return rr.chosenReader.QueryRelationships(ctx, filter, options...)
+}
+
+func (rr *checkingStableReader) ReverseQueryRelationships(
+	ctx context.Context,
+	subjectsFilter datastore.SubjectsFilter,
+	options ...options.ReverseQueryOptionsOption,
+) (datastore.RelationshipIterator, error) {
+	if err := rr.determineSource(ctx); err != nil {
+		return nil, err
+	}
+
+	return rr.chosenReader.ReverseQueryRelationships(ctx, subjectsFilter, options...)
+}
+
+func (rr *checkingStableReader) ReadNamespaceByName(ctx context.Context, nsName string) (ns *core.NamespaceDefinition, lastWritten datastore.Revision, err error) {
+	if err := rr.determineSource(ctx); err != nil {
+		return nil, datastore.NoRevision, err
+	}
+
+	return rr.chosenReader.ReadNamespaceByName(ctx, nsName)
+}
+
+func (rr *checkingStableReader) ListAllNamespaces(ctx context.Context) ([]datastore.RevisionedNamespace, error) {
+	if err := rr.determineSource(ctx); err != nil {
+		return nil, err
+	}
+
+	return rr.chosenReader.ListAllNamespaces(ctx)
+}
+
+func (rr *checkingStableReader) LookupNamespacesWithNames(ctx context.Context, nsNames []string) ([]datastore.RevisionedNamespace, error) {
+	if err := rr.determineSource(ctx); err != nil {
+		return nil, err
+	}
+
+	return rr.chosenReader.LookupNamespacesWithNames(ctx, nsNames)
+}
+
+func (rr *checkingStableReader) CountRelationships(ctx context.Context, filter string) (int, error) {
+	if err := rr.determineSource(ctx); err != nil {
+		return 0, err
+	}
+
+	return rr.chosenReader.CountRelationships(ctx, filter)
+}
+
+func (rr *checkingStableReader) LookupCounters(ctx context.Context) ([]datastore.RelationshipCounter, error) {
+	if err := rr.determineSource(ctx); err != nil {
+		return nil, err
+	}
+
+	return rr.chosenReader.LookupCounters(ctx)
+}
+
+// determineSource will choose the replica or primary to read from based on the revision, by checking
+// if the replica contains the revision. If the replica does not contain the revision, the primary
+// will be used instead.
+func (rr *checkingStableReader) determineSource(ctx context.Context) error {
+	var finalError error
+	rr.choose.Do(func() {
+		// If the revision is not known to the replica, use the primary instead.
+		if err := rr.replica.CheckRevision(ctx, rr.rev); err != nil {
+			var irr datastore.ErrInvalidRevision
```

Comment:
Because they represent different states: one is the revision is invalid (general error returned to callers) vs the specific case where it is unavailable on a replica

Reply:
`ErrInvalidRevision` is a superset. You are using it to determine revision is not replicated. So they are not entirely different states.

Reply:
No, but they are raised by different parts of the code so I want them distinct. `ErrInvalidRevision` could be raised by the replica for a truly invalid revision, and that should *not* fallback to the primary

Reply:
🤷🏻 do not necessarily agree with this, but don't want to block this over it